### PR TITLE
make `Type{Union{}}` `ismutationfree`

### PR DESCRIPTION
**Pull request by gbaraldi**
`Type{Union{}}` and `Core.TypeofBottom` are the same type but not the same object, and `Type{Union{}}` has different flags flags, normalize them so we don't taint functions that return `Union{}`. This was causing exp to not be inacessiblememonly.

It seems this regressed on https://github.com/JuliaLang/julia/pull/48220

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by Keno** [comment]
The ismutationfree flag is supposed to be set for all `Type` instantiations. This seems like a bug.

**Comment by Keno** [comment]
Try this patch:
```
diff --git a/src/jltypes.c b/src/jltypes.c
index 9428bf6a91..d9f50d67d3 100644
--- a/src/jltypes.c
+++ b/src/jltypes.c
@@ -2204,6 +2204,7 @@ void jl_init_types(void) JL_GC_DISABLED
     ((jl_datatype_t*)jl_type_type)->cached_by_hash = 0;
     jl_type_typename->wrapper = jl_new_struct(jl_unionall_type, tttvar, (jl_value_t*)jl_type_type);
     jl_type_type = (jl_unionall_t*)jl_type_typename->wrapper;
+    ((jl_datatype_t*)jl_type_type->body)->ismutationfree = 1;

     jl_typeofbottom_type->super = jl_wrap_Type(jl_bottom_type);

@@ -2807,7 +2808,6 @@ void jl_init_types(void) JL_GC_DISABLED
     jl_symbol_type->ismutationfree = jl_symbol_type->isidentityfree = 1;
     jl_simplevector_type->ismutationfree = jl_simplevector_type->isidentityfree = 1;
     jl_datatype_type->ismutationfree = 1;
-    ((jl_datatype_t*)jl_type_type->body)->ismutationfree = 1;

     // Technically not ismutationfree, but there's a separate system to deal
     // with mutations for global state.
```

**Comment by vtjnash** [comment]
But there is a lot of mutable memory reachable via DataType?

**Comment by Keno** [comment]
We currently have `DataType` be `ismutationfree`, otherwise @gbaraldi's issue would be much larger. If we think that's wrong, because Julia can observe it being mutated, then we need to come up with something better. Note that this predates the introduction of the flag though.

**Comment by vtjnash** [comment]
It seems awkward to need to lie quite so much about it. Unless this predicate (inaccessiblememonly) means that it won't change anything directly through that pointer, which is usually true.

**Comment by gbaraldi** [comment]
Llvm is changing these attributes right now, but it basically means that it doesn't read or write to memory other functions can see.

**Comment by vtjnash** [comment]
`s/see/mutate/`, but that is at odds then with DataType, which generally does reach mutable memory

**Comment by aviatesk** [comment]
Making `!ismutationfree(T::DataType)` would significantly decrease the precision of the `:inaccessiblememonly` analysis. To mitigate this, we could potentially introduce a new bit to `:inaccessiblememonly` to mark access to `DataType` itself as "almost `:inaccessiblememonly`" but taint it immediately when accessing its properties (such as `T.name`).

**Comment by gbaraldi** [comment]
Do you folks want to close this and think of a longer term solution, or merge this as a partial fix and work on something that gives more precise information. I will say that it seems most instances of `Type{whatever}` do have the bit set, and `Union{}` was the odd one out.

**Comment by aviatesk** [comment]
I'd like to go with Keno's patch to fix the precision for now. And then fix the correctness issue in a separate PR.

File: test/compiler/effects.jl [hunk 1]
```diff
@@ -740,3 +740,6 @@ end |> !Core.Compiler.is_consistent
     ImmutRef(x) = $(Expr(:new, :(ImmutRef{typeof(x)}), :x))
 end
 @test Core.Compiler.is_foldable(Base.infer_effects(ImmutRef, Tuple{Any}))
+
+@test Core.Compiler.is_total(Base.infer_effects(typejoin, ()))
+#Test that returning Union{} doesn't taint effects
```
Comment (aviatesk) [inline test/compiler/effects.jl:745]:
```suggestion
@test Base.ismutationfree(Type{Union{}})
@test Core.Compiler.is_total(Base.infer_effects(typejoin, ()))
```

**Pull request merged**